### PR TITLE
Add pruning to bechmarks & update weights.

### PR DESCRIPTION
**Pull request by tomusdrw**
Follow up on #898

### CODE REVIEW

**Review by svyatonik — APPROVED** [review]

File: bin/rialto/runtime/src/lib.rs [hunk 1]
```diff
@@ -414,7 +414,20 @@ parameter_types! {
...
 
+#[cfg(feature = "runtime-benchmarks")]
+parameter_types! {
+	// Number of headers to keep in benchmarks.
+	//
+	// In benchmarks we always populate with full number of `HeadersToKeep` to make sure that
+	// pruning is taken into account.
+	// This is lower than regular value, to speed up benchmarking setup.
+	pub const HeadersToKeep: u32 = 1024;
+}
```
Comment (tomusdrw) [inline bin/rialto/runtime/src/lib.rs:427]: This code is optional, but keeping it reduces the time to run benchmarks ~30% (from 70minutes to 54minutes).

File: modules/grandpa/src/lib.rs [hunk 2]
```diff
@@ -427,6 +415,29 @@ pub mod pallet {
 		)
 	}
 
+	/// Import a previously verified header to the storage.
+	///
+	/// Note this function solely takes care of updating the storage and does not verify
+	/// the validaty of such import.
```
Comment (HCastano) [inline modules/grandpa/src/lib.rs:421]: I would add a note that this function also performs pruning if necessary.

File: modules/grandpa/src/lib.rs [hunk 3]
```diff
@@ -427,6 +415,29 @@ pub mod pallet {
...
+	pub(crate) fn insert_header<T: Config<I>, I: 'static>(
+		header: BridgedHeader<T, I>,
+		hash: Option<BridgedBlockHash<T, I>>,
```
Comment (HCastano) [inline modules/grandpa/src/lib.rs:424]: I don't think it makes sense for this to be an `Option` - it'd just force the caller to provide a `hash`

File: modules/grandpa/src/lib.rs [hunk 4]
```diff
@@ -441,15 +452,38 @@ pub mod pallet {
...
+		while number < end_number {
+			number = number + sp_runtime::traits::One::one();
+			let header = <BridgedHeader<T, I>>::new(
+				number,
+				Default::default(),
+				Default::default(),
+				Default::default(),
+				Default::default(),
+			);
+			insert_header::<T, I>(header, None);
```
Comment (HCastano) [inline modules/grandpa/src/lib.rs:483]:
```suggestion
			insert_header::<T, I>(header, header.hash());
```

File: modules/grandpa/src/lib.rs [hunk 5]
```diff
@@ -441,15 +452,38 @@ pub mod pallet {
...
 		<IsHalted<T, I>>::put(is_halted);
 	}
 
+	#[cfg(feature = "runtime-benchmarks")]
+	pub(crate) fn bootstrap_bridge<T: Config<I>, I: 'static>(
+		init_params: super::InitializationData<BridgedHeader<T, I>>,
+	) {
+		let inject_headers = T::HeadersToKeep::get();
+		let start_number = *init_params.header.number();
+		let end_number = start_number + inject_headers.into();
```
Comment (HCastano) [inline modules/grandpa/src/lib.rs:470]:
```suggestion
		let end_number = start_number + T::HeadersToKeep::get().into();
```

I think this is more readable than `inject_headers`

File: modules/grandpa/src/benchmarking.rs [hunk 6]
```diff
@@ -66,6 +67,10 @@ const MAX_VOTE_ANCESTRIES: u32 = 1000;
 // number of validators.
 const MAX_VALIDATOR_SET_SIZE: u32 = 1024;
 
+fn header_number<T: Config<I>, I: 'static, R: From<u32>>() -> R {
```
Comment (HCastano) [inline modules/grandpa/src/benchmarking.rs:70]: Does `R` stand for something clever here? I would've expected `N` for number, lol

Comment (HCastano) [inline modules/grandpa/src/benchmarking.rs:70]:
```suggestion
// In order to make sure pruning capabilities are being triggered we...
fn header_number<T: Config<I>, I: 'static, R: From<u32>>() -> R {
```

File: bin/rialto/runtime/src/lib.rs [hunk 7]
```diff
@@ -414,7 +414,20 @@ parameter_types! {
...
 
+#[cfg(feature = "runtime-benchmarks")]
+parameter_types! {
+	// Number of headers to keep in benchmarks.
+	//
+	// In benchmarks we always populate with full number of `HeadersToKeep` to make sure that
+	// pruning is taken into account.
+	// This is lower than regular value, to speed up benchmarking setup.
+	pub const HeadersToKeep: u32 = 1024;
```
Comment (HCastano) [inline bin/rialto/runtime/src/lib.rs:426]:
```suggestion
	// pruning is taken into account.
	//
	// Note: This is lower than regular value, to speed up benchmarking setup.
	pub const HeadersToKeep: u32 = 1024;
```

File: bin/rialto/runtime/src/lib.rs (hunk 1 above)
Comment (HCastano) [reply bin/rialto/runtime/src/lib.rs:427]: Is the 70 mins for the full (messages + header) bench suite?

**Review by HCastano — APPROVED** [review]
Some small things but looks good!

File: modules/grandpa/src/benchmarking.rs (hunk 6 above)
Comment (tomusdrw) [reply modules/grandpa/src/benchmarking.rs:70]: it's for "return value", fine to change it to `N` :)

**Comment by tomusdrw** [comment]
> Is the 70 mins for the full (messages + header) bench suite?

Yup, on the standardized hardware :)

**Review by HCastano — APPROVED** [review]

**Comment by HCastano** [comment]
> > Is the 70 mins for the full (messages + header) bench suite?
> 
> Yup, on the standardized hardware :)

Ouf that's rough. Looks like there' still some work to be done on improving the bench times

**Pull request merged**